monai/optimizers/lr_finder.py

```diff
@@ -0,0 +1,629 @@
+                    loss = smooth_f * loss + (1 - smooth_f) * self.history["loss"][-1]
+                if loss < best_loss:
+                    best_loss = loss
+
+            # Check if the loss has diverged; if it has, stop the test
+            self.history["loss"].append(loss)
+            if loss > diverge_th * best_loss:
+                if self.verbose:
+                    tprint("Stopping early, the loss has diverged")
+                break
+
+        if auto_reset:
+            if self.verbose:
+                print("Resetting model and optimizer")
+            self.reset()
+
+    def _set_learning_rate(self, new_lrs: Union[float, list]) -> None:
+        """Set learning rate(s) for optimizer."""
+        if not isinstance(new_lrs, list):
+            new_lrs = [new_lrs] * len(self.optimizer.param_groups)
+        if len(new_lrs) != len(self.optimizer.param_groups):
+            raise ValueError(
+                "Length of `new_lrs` is not equal to the number of parameter groups " + "in the given optimizer"
+            )
+
+        for param_group, new_lr in zip(self.optimizer.param_groups, new_lrs):
+            param_group["lr"] = new_lr
+
+    def _check_for_scheduler(self) -> _none_or_positive_arg:
+        """Check optimizer doesn't already have scheduler."""
+        for param_group in self.optimizer.param_groups:
+            if "initial_lr" in param_group:
+                raise RuntimeError("Optimizer already has a scheduler attached to it")
+
+    def _train_batch(self, train_iter, accumulation_steps: int, non_blocking_transfer: bool = True) -> float:
+        self.model.train()
+        total_loss = 0
+
+        self.optimizer.zero_grad()
+        for i in range(accumulation_steps):
+            inputs, labels = next(train_iter)
+            inputs, labels = self._move_to_device(inputs, labels, non_blocking=non_blocking_transfer)
+
+            # Forward pass
+            outputs = self.model(inputs)
+            loss = self.criterion(outputs, labels)
+
+            # Loss should be averaged in each step
+            loss /= accumulation_steps
+
+            # Backward pass
+            if self.amp and hasattr(self.optimizer, "_amp_stash"):
+                # For minor performance optimization, see also:
+                # https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations
+                delay_unscale = ((i + 1) % accumulation_steps) != 0
+
+                with torch.cuda.amp.scale_loss(loss, self.optimizer, delay_unscale=delay_unscale) as scaled_loss:  # type: ignore
+                    scaled_loss.backward()
+            else:
+                loss.backward()
+
+            total_loss += loss.item()
+
+        self.optimizer.step()
+
+        return total_loss
+
+    def _move_to_device(
+        self, inputs: torch.Tensor, labels: torch.Tensor, non_blocking: bool = True
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        def move(obj, device, non_blocking=True):
```

Comment:
this  could be a generic `to_device` util in monai.utils, but it'll need some recursive logic like this one: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/collate.py#L15